rohmu/atomic_opener.py

```diff
@@ -0,0 +1,128 @@
+from contextlib import contextmanager
+from pathlib import Path
+from typing import BinaryIO, Callable, cast, ContextManager, Generator, Optional, overload, TextIO, TYPE_CHECKING, Union
+
+# workaround < Python3.8 missing Literal.
+# if we just use the try/except block, mypy will complain
+# about the fact that Write* variables are being redefined.
+# so we need to pour in an additional branch.
+if TYPE_CHECKING:
+    from typing import Literal
+
+    Write = Literal["w"]
+    WriteBinary = Literal["wb"]
+    WriteTest = Literal["somethingrandomw"]
+else:
+    try:
```

Comment:
This `try` is probably unecessary. The types can be set to `Any` if it is not type checking

Reply:
Probably. And this will fail if using mypy on Python <= 3.8, which we don't do on our CI.

I personally dislike the Literal typing approach (too complex) but I was asked to do this. 

Is removing the try/except any better, though? I think the else block is a clear, complete path for when we don't use TYPE_CHECKING is unavailable, while TYPE_CHECKING is the exceptional Path that I had to add because of some mypy warts.

Reply:
Possibly, but I really think too much time was wasted on typing here. The original code used a cast, that's true, but it was internal to a module. I think that type checks should help us, not make our life more difficult.

Reply:
> Is removing the try/except any better, though?

Well, it is just a bit less code, but not significatly better. I guess your rationale is "lets try to make the types as close to the correct values as possible", whereas I would go with "lets do the minimum that works for type checking and runtime". No need to change it though :)